### PR TITLE
fix:schema mismatch not detected while executing copy into

### LINKED ISSUE #12684: bug:  schema mismatch not detected while executing `copy into`
**Summary**

version :  [v1.2.98-nightly](https://github.com/datafuselabs/databend/releases/tag/v1.2.98-nightly)

**how to reproduce:**

~~~

mysql> create stage test;
Query OK, 0 rows affected (0.05 sec)

mysql> copy into @test from (select number from numbers(1)) FILE_FORMAT = (type = parquet);
Query OK, 0 rows affected (0.07 sec)

mysql> create table a(c1 int , c2 int);
Query OK, 0 rows affected (0.11 sec)

mysql> copy into a  from (select number  from @test )   file_format = (type = PARQUET);
ERROR 1105 (HY000): PanicError. Code: 1104, Text = index out of bounds: the len is 1 but the index is 1.

~~~

log shows that statement exec panicked druing constructing bloom index

~~~
panicked at 'index out of bounds: the len is 1 but the index is 1', ~/workspace/fuse-query/src/query/expression/src/block.rs:188:10
   0: common_tracing::panic_hook::log_panic
             at ./src/common/tracing/src/panic_hook.rs:35:21
   1: common_tracing::panic_hook::set_panic_hook::{{closure}}
             at ./src/common/tracing/src/panic_hook.rs:30:9
   2: call<(&core::panic::panic_info::PanicInfo), (dyn core::ops::function::Fn<(&core::panic::panic_info::PanicInfo), Output=()> + core::marker::Send + core::marker::Sync), alloc::alloc::Global>
             at /rustc/8b35c0bb0f833c0077dc57006eb317edde2a2d1e/library/alloc/src/boxed.rs:1999:9
   3: rust_panic_with_hook
             at /rustc/8b35c0bb0f833c0077dc57006eb317edde2a2d1e/library/std/src/panicking.rs:709:13
   4: {closure#0}
             at /rustc/8b35c0bb0f833c0077dc57006eb317edde2a2d1e/library/std/src/panicking.rs:597:13
   5: __rust_end_short_backtrace<std::panicking::begin_panic_handler::{closure_env#0}, !>
             at /rustc/8b35c0bb0f833c0077dc57006eb317edde2a2d1e/library/std/src/sys_common/backtrace.rs:151:18
   6: begin_panic_handler
             at /rustc/8b35c0bb0f833c0077dc57006eb317edde2a2d1e/library/std/src/panicking.rs:593:5
   7: panic_fmt
             at /rustc/8b35c0bb0f833c0077dc57006eb317edde2a2d1e/library/core/src/panicking.rs:67:14
   8: panic_bounds_check
             at /rustc/8b35c0bb0f833c0077dc57006eb317edde2a2d1e/library/core/src/panicking.rs:162:5
   9: <usize as core::slice::index::SliceIndex<[T]>>::index
             at /rustc/8b35c0bb0f833c0077dc57006eb317edde2a2d1e/library/core/src/slice/index.rs:258:10
  10: <alloc::vec::Vec<T,A> as core::ops::index::Index<I>>::index
             at /rustc/8b35c0bb0f833c0077dc57006eb317edde2a2d1e/library/core/src/slice/index.rs:18:9
  11: common_expression::block::DataBlock::get_by_offset
             at ./src/query/expression/src/block.rs:188:10
  12: storages_common_index::bloom_index::BloomIndex::try_create
             at ./src/query/storages/common/index/src/bloom_index.rs:194:31
  13: common_storages_fuse::io::write::block_writer::BloomIndexState::try_create
             at ./src/query/storages/fuse/src/io/write/block_writer.rs:123:33
  14: common_storages_fuse::io::write::block_writer::BlockBuilder::build
             at ./src/query/storages/fuse/src/io/write/block_writer.rs:178:33
  15: <common_storages_fuse::operations::common::processors::transform_serialize_block::TransformSerializeBlock as common_pipeline_core::processors::processor::Processor>::process
             at ./src/query/storages/fuse/src/operations/common/processors/transform_serialize_block.rs:219:21
  16: <alloc::boxed::Box<T> as common_pipeline_core::processors::processor::Processor>::process
             at ./src/query/pipeline/core/src/processors/processor.rs:138:9
  17: common_pipeline_core::processors::processor::ProcessorPtr::process
             at ./src/query/pipeline/core/src/processors/processor.rs:110:9
  18: databend_query::pipelines::executor::executor_worker_context::ExecutorWorkerContext::execute_sync_task
             at ./src/query/service/src/pipelines/executor/executor_worker_context.rs:82:9
  19: databend_query::pipelines::executor::executor_worker_context::ExecutorWorkerContext::execute_task
             at ./src/query/service/src/pipelines/executor/executor_worker_context.rs:73:46
  20: databend_query::pipelines::executor::pipeline_executor::PipelineExecutor::execute_single_thread
             at ./src/query/service/src/pipelines/executor/pipeline_executor.rs:376:45
  21: databend_query::pipelines::executor::pipeline_executor::PipelineExecutor::execute_threads::{{closure}}::{{closure}}
             at ./src/query/service/src/pipelines/executor/pipeline_executor.rs:332:27
  22: <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
             at /rustc/8b35c0bb0f833c0077dc57006eb317edde2a2d1e/library/core/src/panic/unwind_safe.rs:271:9
  23: std::panicking::try::do_call
             at /rustc/8b35c0bb0f833c0077dc57006eb317edde2a2d1e/library/std/src/panicking.rs:500:40
  24: __rust_try
  25: std::panicking::try
             at /rustc/8b35c0bb0f833c0077dc57006eb317edde2a2d1e/library/std/src/panicking.rs:464:19
  26: std::panic::catch_unwind
             at /rustc/8b35c0bb0f833c0077dc57006eb317edde2a2d1e/library/std/src/panic.rs:142:14
  27: common_base::runtime::catch_unwind::catch_unwind
             at ./src/common/base/src/runtime/catch_unwind.rs:26:11
  28: databend_query::pipelines::executor::pipeline_executor::PipelineExecutor::execute_threads::{{closure}}
             at ./src/query/service/src/pipelines/executor/pipeline_executor.rs:331:34
  29: common_base::runtime::thread::Thread::named_spawn::{{closure}}
             at ./src/common/base/src/runtime/thread.rs:85:21
  30: std::sys_common::backtrace::__rust_begin_short_backtrace
             at /rustc/8b35c0bb0f833c0077dc57006eb317edde2a2d1e/library/std/src/sys_common/backtrace.rs:135:18
  31: std::thread::Builder::spawn_unchecked_::{{closure}}::{{closure}}
             at /rustc/8b35c0bb0f833c0077dc57006eb317edde2a2d1e/library/std/src/thread/mod.rs:529:17
  32: <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
             at /rustc/8b35c0bb0f833c0077dc57006eb317edde2a2d1e/library/core/src/panic/unwind_safe.rs:271:9
  33: std::panicking::try::do_call
             at /rustc/8b35c0bb0f833c0077dc57006eb317edde2a2d1e/library/std/src/panicking.rs:500:40
  34: __rust_try
  35: std::panicking::try
             at /rustc/8b35c0bb0f833c0077dc57006eb317edde2a2d1e/library/std/src/panicking.rs:464:19
  36: std::thread::Builder::spawn_unchecked_::{{closure}}
             at /rustc/8b35c0bb0f833c0077dc57006eb317edde2a2d1e/library/std/src/panic.rs:142:14
  37: core::ops::function::FnOnce::call_once{{vtable.shim}}
             at /rustc/8b35c0bb0f833c0077dc57006eb317edde2a2d1e/library/core/src/ops/function.rs:250:5
  38: call_once<(), dyn core::ops::function::FnOnce<(), Output=()>, alloc::alloc::Global>
             at /rustc/8b35c0bb0f833c0077dc57006eb317edde2a2d1e/library/alloc/src/boxed.rs:1985:9
  39: call_once<(), alloc::boxed::Box<dyn core::ops::function::FnOnce<(), Output=()>, alloc::alloc::Global>, alloc::alloc::Global>
             at /rustc/8b35c0bb0f833c0077dc57006eb317edde2a2d1e/library/alloc/src/boxed.rs:1985:9
  40: thread_start
             at /rustc/8b35c0bb0f833c0077dc57006eb317edde2a2d1e/library/std/src/sys/unix/thread.rs:108:17
  41: start_thread
             at ./nptl/pthread_create.c:442:8
  42: clone3
             at ./misc/../sysdeps/unix/sysv/linux/x86_64/clone3.S:81
~~~

**what we expect:**

sth like this will be better

~~~
mysql> insert into a select number  from @test;
ERROR 1105 (HY000): BadArguments. Code: 1006, Text = Fields in select statement is not equal with expected, select fields: 1, insert fields: 2.
mysql>
~~~

### PR DESCRIPTION
**Pull request by zenus**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

Summary about this PR

- Closes #12684

- - -
This change is [Reviewable](https://reviewable.io/reviews/datafuselabs/databend/13010)

### CODE REVIEW

**Comment by BohuTANG** [comment]
I think this is fixed in this PR: https://github.com/datafuselabs/databend/pull/12990

File: src/query/sql/src/planner/binder/copy.rs [hunk 1]
```diff
@@ -213,7 +213,13 @@ impl<'a> Binder {
                     }
                     .into(),
                 );
-
+                if required_values_schema.fields().len() != select_list.len() {
+                    return Err(ErrorCode::BadArguments(format!(
+                        "Fields in select statement is not equal with expected, select fields: {}, insert fields: {}",
+                        select_list.len(),
+                        required_values_schema.fields().len(),
```
Comment (youngsofun) [inline src/query/sql/src/planner/binder/copy.rs:220]:
```suggestion
                        "Number of columns in select list {} does not match that of the corresponding table {}",
                        select_list.len(),
                        required_values_schema.fields().len(),
```

Comment (youngsofun) [reply src/query/sql/src/planner/binder/copy.rs:220]: here I suggest using similar words with 
https://github.com/datafuselabs/databend/pull/12888

**Comment by BohuTANG** [comment]
> I think this is fixed in this PR: #12990

It seems not fixed but the error is not same as #12990 

Thanks @zenus , let's continue

File: tests/sqllogictests/suites/base/03_common/03_0028_copy_into_stage [hunk 2]
```diff
@@ -37,4 +37,26 @@ statement ok
...
+statement ok
+CREATE STAGE IF NOT EXISTS hello
+
+statement ok
+COPY INTO @hello from (select number from numbers(1)) FILE_FORMAT = (type = parquet)
+
+statement ok
+CREATE TABLE word(c1 int , c2 int);
+
+statement error (?s)1006.* Fields in select statement is not equal with expected, select fields: 1, insert fields: 2
```
Comment (youngsofun) [inline tests/sqllogictests/suites/base/03_common/03_0028_copy_into_stage:51]:
```suggestion
statement error (?s)1006.* Number of columns in select list (1) does not match that of the corresponding table (2)"
```

**Comment by youngsofun** [comment]
```
0: query failed: mysql client error: Server error: `ERROR HY000 (1105): BadArguments. Code: 1006, Text = Fields in select statement is not equal with expected, select fields: 1, insert fields: 2.'
[SQL] copy /*+ set_var(parquet_fast_read_bytes=0) */ into ai from ( select * from  @data/parquet/multi_page/ )  PATTERN='.*[.]parquet' force=true;
at tests/sqllogictests/suites/stage_parquet/optimize_small_file:36
```

seems you need to handle `select *`

File: src/query/sql/src/planner/binder/copy.rs (hunk 1 above)
Comment (zenus) [reply src/query/sql/src/planner/binder/copy.rs]: done

File: tests/sqllogictests/suites/base/03_common/03_0028_copy_into_stage (hunk 2 above)
Comment (zenus) [reply tests/sqllogictests/suites/base/03_common/03_0028_copy_into_stage]: done

**Comment by zenus** [comment]
> ```
> 0: query failed: mysql client error: Server error: `ERROR HY000 (1105): BadArguments. Code: 1006, Text = Fields in select statement is not equal with expected, select fields: 1, insert fields: 2.'
> [SQL] copy /*+ set_var(parquet_fast_read_bytes=0) */ into ai from ( select * from  @data/parquet/multi_page/ )  PATTERN='.*[.]parquet' force=true;
> at tests/sqllogictests/suites/stage_parquet/optimize_small_file:36
> ```
> 
> seems you need to handle `select *`
> 
> btw: it should be `select fields: 2, insert fields: 1`

done

File: tests/sqllogictests/suites/base/03_common/03_0028_copy_into_stage (hunk 2 above)
Comment (zenus) [reply tests/sqllogictests/suites/base/03_common/03_0028_copy_into_stage]:
```rust
Test finished, fail fast enabled, 1 out of 4407 records failed to run
0: statement is expected to fail with error:
	(?s)1006.* Number of columns in select list (1) does not match that of the corresponding table (2)
but got error:
Error: SelfError("sqllogictest failed")
	mysql client error: Server error: `ERROR HY000 (1105): BadArguments. Code: 1006, Text = Number of columns in select list (1) does not match that of the corresponding table (2).'
[SQL] COPY INTO world FROM (select number  from @hello )   file_format = (type = PARQUET);
at tests/sqllogictests/suites/base/03_common/03_0028_copy_into_stage:51
``` 
confused

File: tests/sqllogictests/suites/base/03_common/03_0028_copy_into_stage [hunk 3]
```diff
@@ -37,4 +37,26 @@ statement ok
...
+statement ok
+CREATE STAGE IF NOT EXISTS hello
+
+statement ok
+COPY INTO @hello from (select number from numbers(1)) FILE_FORMAT = (type = parquet)
+
+statement ok
+CREATE TABLE world(c1 int , c2 int);
+
+statement error (?s)1006.* Number of columns in select list (1) does not match that of the corresponding table (2)
```
Comment (youngsofun) [inline tests/sqllogictests/suites/base/03_common/03_0028_copy_into_stage:51]:
```suggestion
query error (?s)1006.* Number of columns in select list (1) does not match that of the corresponding table (2)
```

File: tests/sqllogictests/suites/base/03_common/03_0028_copy_into_stage (hunk 2 above)
Comment (youngsofun) [reply tests/sqllogictests/suites/base/03_common/03_0028_copy_into_stage]: let me try `statement` -> `query`

File: tests/sqllogictests/suites/base/03_common/03_0028_copy_into_stage [hunk 4]
```diff
@@ -37,4 +37,26 @@ statement ok
...
+statement ok
+CREATE STAGE IF NOT EXISTS hello
+
+statement ok
+COPY INTO @hello from (select number from numbers(1)) FILE_FORMAT = (type = parquet)
+
+statement ok
+CREATE TABLE world(c1 int , c2 int);
+
+query error (?s)1006.* Number of columns in select list (1) does not match that of the corresponding table (2)
```
Comment (youngsofun) [inline tests/sqllogictests/suites/base/03_common/03_0028_copy_into_stage:51]:
```suggestion
statement error (?s)1006.*Number of columns in select list \(1\) does not match that of the corresponding table \(2\)
```

File: tests/sqllogictests/suites/base/03_common/03_0028_copy_into_stage (hunk 2 above)
Comment (youngsofun) [reply tests/sqllogictests/suites/base/03_common/03_0028_copy_into_stage]: I got it, it`s regex, need to escape `(` and `)`

**Comment by zenus** [comment]
ok，let&nbsp; me fix ,and add more&nbsp; case.



---Original---
From: "Yang ***@***.***&gt;
Date: Thu, Sep 28, 2023 20:19 PM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [datafuselabs/databend] fix:schema mismatch not detected whileexecuting copy into (PR #13010)




 
@youngsofun commented on this pull request.
 
 
In tests/sqllogictests/suites/base/03_common/03_0028_copy_into_stage:
 &gt; \ No newline at end of file +DROP STAGE IF EXISTS hello + +statement ok +CREATE STAGE IF NOT EXISTS hello + +statement ok +COPY INTO @hello from (select number from numbers(1)) FILE_FORMAT = (type = parquet) + +statement ok +CREATE TABLE word(c1 int , c2 int); + +statement error (?s)1006.* Fields in select statement is not equal with expected, select fields: 1, insert fields: 2  
I got it, its regex, need to escape (and)`

**Review by youngsofun — APPROVED** [review]

**Pull request merged**